### PR TITLE
Fixes #4193 - issue with the exclude/include for InetAccesHandler - add better unit test

### LINKED ISSUE #4193: InetAccessHandler - new includeConnectors/excludeConnectors not quite correct anymore
In the original PR i opened... https://github.com/eclipse/jetty.project/pull/3572

We had the logic correct at this point.

Then the PR was taken and extended here: https://github.com/eclipse/jetty.project/pull/3576

Here we lost the key functionality that I added during this PR - we can no longer exclude some connectors from the InetAccessHandler, and include others. the logic was incorrectly applied as exclusion rule on the connector name itself, was was not correct. 

It was my fault because my unit test was horrible and didn't test the new feature at all.

### PR DESCRIPTION
**Pull request by nddipiazza**
fixes https://github.com/eclipse/jetty.project/issues/4193

add a much more in depth unit test to cover this scenario.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java [hunk 1]
```diff
@@ -188,13 +188,16 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
     protected boolean isAllowed(InetAddress addr, Request baseRequest, HttpServletRequest request)
     {
         String name = baseRequest.getHttpChannel().getConnector().getName();
+        boolean includedConnectorName = _names.isIncluded(name);
+        boolean excludedConnectorName = _names.isExcluded(name);
+        boolean filterAppliesToThisConnector = (_names.includesSize() == 0 || includedConnectorName) && !excludedConnectorName;
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java]: The `_names.isIncludedAndNotExcluded(name);` from below is what this block of code does.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java [hunk 2]
```diff
@@ -188,13 +188,16 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
         if (LOG.isDebugEnabled())
         {
-            Boolean allowedByName = _names.isIncludedAndNotExcluded(name);
             Boolean allowedByAddr = _addrs.isIncludedAndNotExcluded(addr);
-            LOG.debug("{} allowedByName={} allowedByAddr={} for {}/{}", this, allowedByName, allowedByAddr, addr, request);
+            LOG.debug("{} includedConnectorName={} excludedConnectorName={} filterAppliesToThisConnector={} allowedByAddr={} for {}/{}",
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java]: leave the  "allowedBy____" syntax.
We don't want to get in the habit of dumping every little include/exclude for every possible path through this.
That's what the `.dump()` methods are for.

File: jetty-server/src/test/java/org/eclipse/jetty/server/handler/InetAccessHandlerTest.java [hunk 3]
```diff
@@ -136,39 +155,62 @@ public void testHandler(String include, String exclude, String includeConnectors
...
+                {"127.0.0.1", "", "", "http_connector1;http_connector2", "200;200"},
+                {"127.0.0.1-127.0.0.254", "", "", "http_connector1;http_connector2", "200;200"},
+                {"192.0.0.1", "", "", "http_connector1;http_connector2", "200;200"},
+                {"192.0.0.1-192.0.0.254", "", "", "http_connector1;http_connector2", "200;200"},
+
+                // 26 - 29 - both connectors are included
+                {"127.0.0.1", "", "http_connector1;http_connector2", "", "200;200"},
+                {"127.0.0.1-127.0.0.254", "", "http_connector1;http_connector2", "", "200;200"},
+                {"192.0.0.1", "", "http_connector1;http_connector2", "", "403;403"},
+                {"192.0.0.1-192.0.0.254", "", "http_connector1;http_connector2", "", "403;403"},
```
Comment (joakime) [inline jetty-server/src/test/java/org/eclipse/jetty/server/handler/InetAccessHandlerTest.java]: +1 for the improved testcase.

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 4]
```diff
@@ -188,6 +188,28 @@ public Boolean isIncludedAndNotExcluded(P item)
...
+     * Test Excluded
+     *
+     * @param item The item to test
+     * @return Boolean.TRUE if item is excluded
+     */
+    public Boolean isExcluded(P item)
+    {
+        return _excludePredicate.test(item);
+    }
+
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: These are not supposed to be here.
This is a Predicate that follows the "IncludeExcludeSet" logic.
Where Excludes always win over Includes.
Please remove these.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java (hunk 2 above)
Comment (nddipiazza) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java]: Hi. I went ahead and dumped in something that definitely works for now. Will get back to the cleaner version once i figure out why it doesn't work.

**Comment by nddipiazza** [comment]
Hey @joakime I put back the original logic before my changes. 

The unit test will now fail because the logic isn't quite what it should be. 

CC @gregw  

in particular, this logic:

```
        String name = baseRequest.getHttpChannel().getConnector().getName();       
        return _names.test(name) && _addrs.test(addr);
```

Is not correct. it's treating the connector name exactly like the filter. But that's not what it's intended to do. It's supposed to tell what connectors are applicable to this filter. And what connectors are not affected.

For example in the unit test there exists 2 connectors:

`http`
`tls`

We want to restrict the `http` connector, but we want to leave `tls` connector alone.

So we would specify:

`include` = `192.168.1.1-192.168.1.254`
`includeConnector` = `http`

The way the logic is above, it is treating the connector name as if it's the filter itself. Which is not what I intended.

What i need in psuedo-code is this:

```

   if (there are no "include connectors" OR if this connector is included) AND (if this connector is not in the excluded list)
     ---> Then apply the IP filter.
```

I can't see any way to easily accomplish this with the `IncludeExcludeSet` ?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java (hunk 2 above)
Comment (nddipiazza) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/InetAccessHandler.java]: OK i made some changes that work for me that still use this `IncludeExcludeSet` class. please review.

**Comment by nddipiazza** [comment]
@joakime ok i made the change you requested about the "exclude should precede the include".

**Comment by gregw** [comment]
@nddipiazza 
I think we should add the following utility method to `IncludeExcludeSet`:
```java

    /** 
     * Match items in combined IncludeExcludeSets. 
     * @param item1 The item to match against set1
     * @param set1 A IncludeExcludeSet to match item1 against
     * @param item2 The item to match against set2
     * @param set2 A IncludeExcludeSet to match item2 against
     * @param <T1> The type of item1
     * @param <T2> The type of item2
     * @return True IFF <ul>
     *     <li>Neither item is excluded from their respective sets</li>
     *     <li>Both sets have no includes OR at least one of the items is included in its respective set</li>
     * </ul>
     */
    public static <T1,T2> boolean matchCombined(T1 item1, IncludeExcludeSet<?,T1> set1, T2 item2, IncludeExcludeSet<?,T2> set2)
    {
        Boolean match1 = set1.isIncludedAndNotExcluded(item1);
        Boolean match2 = set2.isIncludedAndNotExcluded(item2);
        
        // if we are excluded from either set, then we do not match
        if (match1 == Boolean.FALSE || match2 == Boolean.FALSE)
            return false;
        
        // If either set has any includes, then we must be included by one of them
        if (set1.hasIncludes() || set2.hasIncludes())
            return match1 == Boolean.TRUE || match2 == Boolean.TRUE;
        
        // If not excluded and no includes, then we match
        return true;
    }
```

The method in the InetAccessHandler then just becomes:
```java
    protected boolean isAllowed(InetAddress addr, Request baseRequest, HttpServletRequest request)
    {
        String name = baseRequest.getHttpChannel().getConnector().getName();
        return IncludeExcludeSet.matchCombined(addr,_addrs,name,_names);
    }
```

I hope this passes your updated test harness?!?!?!
Note that we are trying to get a release rolled tomorrow, so if you could update/test/submit this before then, that would be great!

**Comment by nddipiazza** [comment]
@gregw  No this doesn't pass the test suite. it breaks all of the conditionals like what we have in jetty-9.4.x right now.

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 5]
```diff
@@ -180,12 +180,10 @@ public boolean test(P t)
...
     {
-        if (_excludePredicate.test(item))
+        if (_excludePredicate.test(item)) {
             return Boolean.FALSE;
-        if (_includePredicate.test(item))
-            return Boolean.TRUE;
-
-        return null;
+        }
+        return _includes.size() == 0 || _includePredicate.test(item);
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: This is definitely a wrong change!  The point of this method is that it is a try state - sometimes we depend on the null return

**Comment by gregw** [comment]
@nddipiazza I've pulled your changes into #4206 and cleaned up a few things, so can you review that and I'm closing this one.

**Pull request closed without merging**